### PR TITLE
Run unit tests based on msgpack_test_fixtures

### PR DESCRIPTION
**Pull request by amnonbc**
Run unit tests using fixtures from  https://github.com/ably/ably-common/blob/main/test-resources/messages-encoding.json and https://github.com/ably/ably-common/blob/main/test-resources/msgpack_test_fixtures.json
Ensure that the json and msgpack messages are decoded correctly, and that re-encoding the decoded message 
gets you back to the original fixture.

### CODE REVIEW

File: ably/proto_message_decoding_test.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/ably/ably-go/ably/internal/ablyutil"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
```
Comment (lmars) [inline ably/proto_message_decoding_test.go:13]: Should we use either `assert` or `require` rather than a mixture of both?

File: ably/proto_message_decoding_test.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+)
+
+type fixture struct {
+	Data          string `json:"data"`
+	Encoding      string `json:"encoding"`
+	ExpectedType  string `json:"expectedType"`
+	ExpectedValue any    `json:"expectedValue"`
+}
+
+func loadFixtures() ([]fixture, error) {
```
Comment (lmars) [inline ably/proto_message_decoding_test.go:23]: There is some similar code for loading the crypto fixtures in the `ablytest` package [here](https://github.com/ably/ably-go/blob/main/ablytest/cryptodata.go), is it worth being consistent with that existing code?

File: ably/proto_message_decoding_test.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+	err = json.Unmarshal(text, &dec)
+	return dec.Messages, err
+}
+
+func Test_decodeMessage(t *testing.T) {
+	fixtures, err := loadFixtures()
+	require.NoError(t, err, "failed to load test fixtures")
+	for _, f := range fixtures {
+		t.Run(f.Data, func(t *testing.T) {
```
Comment (lmars) [inline ably/proto_message_decoding_test.go:42]: Is it worth giving each fixture a human readable name rather than using the data as the test name?

File: ably/proto_message_decoding_test.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+			case "jsonObject":
+				assert.IsType(t, map[string]any{}, decodedMsg.Data)
+				assert.Equal(t, f.ExpectedValue, decodedMsg.Data)
+
+			case "jsonArray":
+				assert.IsType(t, []any{}, decodedMsg.Data)
+				assert.Equal(t, f.ExpectedValue, decodedMsg.Data)
+			case "binary":
+				assert.IsType(t, []byte{}, decodedMsg.Data)
+				if f.ExpectedValue == nil {
```
Comment (lmars) [inline ably/proto_message_decoding_test.go:62]: Should all fixtures not have an expected value?

File: go.mod [hunk 5]
```diff
@@ -1,5 +1,7 @@
 module github.com/ably/ably-go
 
+go 1.18
+
```
Comment (lmars) [inline go.mod:4]: Did you intend to include this change? It seems unrelated

File: ably/proto_message_decoding_test.go (hunk 1 above)
Comment (amnonbc) [reply ably/proto_message_decoding_test.go:13]: I am all for reducing dependencies. 
But assert and require do different things.
We use assert to record an error, but allow the test to continue to run, so other failures can be found.
On the other hand use require for things like failure to open the fixture file, where there is no way to continue the error and we need to abort the test now.

File: ably/proto_message_decoding_test.go (hunk 2 above)
Comment (amnonbc) [reply ably/proto_message_decoding_test.go:23]: Both functions read test fixture data. But the data does have different schemas, and different encodings.
And LoadCryptoData() returns different data types. 
DNRY is a good principle. But here, a I think in this case, and in loadMsgpackFixtures, a bit of repetition leads to simpler and more understandable code

File: ably/proto_message_decoding_test.go (hunk 3 above)
Comment (amnonbc) [reply ably/proto_message_decoding_test.go:42]: That is what I did in the msgpack fixtures.
But the messages-encoding.json file was created by [many years ago](https://github.com/ably/ably-common/pull/17) and does not have a name field. The fixtures in this file were not used in any tests in ably-go, so I decided to add one. But I did not intend to alter this file, which probably is imported by other SDKs.

File: ably/proto_message_decoding_test.go (hunk 4 above)
Comment (amnonbc) [reply ably/proto_message_decoding_test.go:62]: if they have a legal decoding, yes
But "3q2+7w==" (which is base64 for 0xdeadbeef) does not have an ExpectedValue in messages-encoding.json
(though it does have an expectedHexValue).

File: go.mod (hunk 5 above)
Comment (amnonbc) [reply go.mod:4]: The new tests won't compile without this because it includes the `any` type, which was introduced in 1.18.
If you think it is important, I could change all occurrences of `any` to `[]byte`

File: ably/proto_message_decoding_test.go (hunk 1 above)
Comment (lmars) [reply ably/proto_message_decoding_test.go:13]: Ok thanks, so let me instead comment in the code, since on a first pass I was under the impression this was not how they were being used and you meant to use `require`.

File: ably/proto_message_decoding_test.go (hunk 2 above)
Comment (lmars) [reply ably/proto_message_decoding_test.go:23]: Ok sorry, I didn't mean use that exact method, but perhaps have `ablytest.LoadMessageEncodingFixtures` and `ablytest.LoadMsgpackFixtures` (i.e. put all the test fixture loading code in the same place), but I understand if you don't think that's useful.

File: ably/proto_message_decoding_test.go (hunk 3 above)
Comment (lmars) [reply ably/proto_message_decoding_test.go:42]: Ok I was implying we add a name to the fixture file, but I'll leave it to your judgement, happy to also leave as is.

File: ably/proto_message_decoding_test.go (hunk 4 above)
Comment (lmars) [reply ably/proto_message_decoding_test.go:62]: So should this be checking `expectedHexValue`?

File: go.mod (hunk 5 above)
Comment (lmars) [reply go.mod:4]: Ok thanks, so I think this should be in a separate commit with that context in the commit message.

File: ably/proto_message_decoding_test.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	fixtures, err := loadFixtures()
+	require.NoError(t, err, "failed to load test fixtures")
+	for _, f := range fixtures {
+		t.Run(f.Data, func(t *testing.T) {
+			msg := Message{
+				Data:     f.Data,
+				Encoding: f.Encoding,
+			}
+			decodedMsg, err := msg.withDecodedData(nil)
+			assert.NoError(t, err)
```
Comment (lmars) [inline ably/proto_message_decoding_test.go:49]: I think this should be `require.NoError` since any further assertions seem meaningless as `decodedMsg` is a zero value.

File: go.mod (hunk 5 above)
Comment (amnonbc) [reply go.mod:4]: I will squash merge this anyway, so the commits will be mushed together into one commit.
Should I do a separate PR for this?

File: ably/proto_message_decoding_test.go (hunk 4 above)
Comment (amnonbc) [reply ably/proto_message_decoding_test.go]: done in https://github.com/ably/ably-go/pull/588/commits/707c94f5163d8d1422cc6efbd0a11e6b102094d5

File: ably/proto_message_decoding_test.go (hunk 3 above)
Comment (amnonbc) [reply ably/proto_message_decoding_test.go:43]: I don't really want to change  messages-encoding.json as it is being imported by other SDKs

File: ably/proto_message_decoding_test.go (hunk 2 above)
Comment (amnonbc) [reply ably/proto_message_decoding_test.go:24]: I could do that if you want. But it is a trivial function, so I think it is probably better to have it not exported and defined nearer the test that uses it.

Comment (lmars) [reply ably/proto_message_decoding_test.go:24]: 👍

File: ably/proto_message_decoding_test.go (hunk 3 above)
Comment (lmars) [reply ably/proto_message_decoding_test.go:43]: 👍

File: ably/proto_message_decoding_test.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type fixture struct {
+	Data          string `json:"data"`
+	Encoding      string `json:"encoding"`
+	ExpectedType  string `json:"expectedType"`
+	ExpectedValue any    `json:"expectedValue"`
+	ExpectedHex   []byte `json:"expectedHex"`
```
Comment (lmars) [inline ably/proto_message_decoding_test.go:21]: I thought it was called `expectedHexValue`?

File: go.mod (hunk 5 above)
Comment (lmars) [reply go.mod:4]: No separate PR no, just merge with two commits?

File: ably/proto_message_decoding_test.go (hunk 7 above)
Comment (amnonbc) [reply ably/proto_message_decoding_test.go]: well spotted.
Fixed

**Review by lmars — APPROVED** [review]

**Review by AndyTWF — APPROVED** [review]
